ipn/localapi/localapi.go

```diff
@@ -730,6 +743,50 @@ func (h *Handler) serveDial(w http.ResponseWriter, r *http.Request) {
 	<-errc
 }
 
+func (h *Handler) serveUploadClientMetrics(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "unsupported method", http.StatusMethodNotAllowed)
+		return
+	}
+	type clientMetricJSON struct {
+		Name string `json:"name"`
+		// One of "counter" or "gauge"
+		Type  string `json:"type"`
+		Value int    `json:"value"`
+	}
+
+	var clientMetrics []clientMetricJSON
+	if err := json.NewDecoder(r.Body).Decode(&clientMetrics); err != nil {
+		http.Error(w, "invalid JSON body", 400)
+		return
+	}
+
+	metricsMu.Lock()
+	defer metricsMu.Unlock()
+
+	for _, m := range clientMetrics {
+		if metric, ok := metrics[m.Name]; ok {
+			metric.Add(int64(m.Value))
+		} else {
+			var metric *clientmetric.Metric
+			switch m.Type {
+			case "counter":
+				metric = clientmetric.NewCounter(m.Name)
+			case "gauge":
+				metric = clientmetric.NewGauge(m.Name)
+			default:
```

Comment:
These will panic if a localapi client provides an m.Name that's already in use, though.

I guess inside an HTTP handler that'll get caught and it'll be an HTTP error at least?

Either add a comment saying that's acceptable, or maybe add new api to the clientmetric package to test whether a metric name is already published?

Reply:
Yeah, the panics were being recovered (though I don't actually see where that's happening), but it doesn't seem ideal to rely on that.

Added a `HasPublished` API and made the API call return a 400 in that case.